insights/parsers/__init__.py

```diff
@@ -498,38 +506,90 @@ def keyword_search(rows, **kwargs):
         'endswith': lambda s, v: s is not None and s.endswith(v),
         'lower_value': lambda s, v: None not in (s, v) and s.lower() == v.lower(),
     }
 
-    def key_match(row, key, value):
-        # Translate ' ' and '-' of keys in dict to '_' to match keyword arguments.
-        my_row = {}
-        for my_key, val in row.items():
-            my_row[my_key.replace(' ', '_').replace('-', '_')] = val
-        matcher_fn = matchers['default']
-        if '__' in key:
-            key, matcher = key.split('__', 1)
+    txform_cache_attr = '_transform_cache'
+    if parent is None and hasattr(rows, '__dict__'):
+        parent = rows
+    # Uncomment this 'if' to check that all the parsers are supplying an
+    # object, somehow, that can store our transformed row cache.
+    # if parent is None:
+    #     print("Invoked with no parent arg on primitive container - use parent=self argument to cache row transform")
+    # The actual txkeys cache gets used rarely - in a basic test with a real
+    # archive it only got used eleven times.  But that's a saving...
+    if parent is not None and hasattr(parent, txform_cache_attr):
+        txkeys = getattr(parent, txform_cache_attr)
+    else:
+        # Store the translation from the search key to the key in the data.
+        # We start with the reverse transform - the key, to the transformed
+        # key.  This avoids doing the string replacement for every key in
+        # every row.  Then we invert that to map from the search term (with
+        # spaces and dashes replaced by underscores) to the key in the row
+        # data.
+        rev_txkeys = dict()
+        for row in rows:
+            for key in row.keys():
+                if key not in rev_txkeys:
+                    # Transform the term in a keyword_search() call into the
+                    # key in the rows
+                    rev_txkeys[key] = key.replace(' ', '_').replace('-', '_')
+        txkeys = dict(
+            (txkey, key)
+            for key, txkey in rev_txkeys.items()
+        )
+        if parent is not None:
+            setattr(parent, txform_cache_attr, txkeys)
+
+    # pre-compile the kwargs to find the matcher function and underlying key.
+    # Store these in a list of tuples for fast iteration and unpacking
+    search_terms = list()
+    for search_keyword, value in kwargs.items():
+        data_key = search_keyword
+        if '__' not in data_key:
+            matcher = 'equals'
+        else:
+            data_key, matcher = search_keyword.split('__', 1)
             if matcher not in matchers:
                 # put key back the way we found it, matcher fn unchanged
-                key = key + '__' + matcher
-            else:
-                matcher_fn = matchers[matcher]
-        return key in my_row and matcher_fn(my_row[key], value)
-
-    data = []
+                data_key = search_keyword
+                matcher = 'equals'
```

Comment:
Here, your code definitely looks good, this is just kind of nitpick.
How about using the `partition()` to simplify the code as:
```python
for search_keyword, value in kwargs.items():
   data_key, _, matcher = search_keyword.partition('__')
   if matcher not in matchers:
        matcher = 'equals'
   ...
```

Reply:
Hmmm.  Interesting approach!  It's faster too AFAICS, so that's what we'll do.

Reply:
The thing we're trying to handle is the weird case that a column might be named 'foo__bar', which looks like a 'foo__startswith' matcher invocation.  I can see the use of the `contains` and `startswith` and even `endswith` in the plugins, so these are being used; I can also see plenty of uses of search without the matcher invocation.  I can't see any uses of `__` in a keyword that isn't followed by a matcher.  So we want to optimise for searches with no '__', then finding the matcher as quickly as possible, and then finally the slow path of a keyword that has '__' in it but no matcher.